onmt/Loss.py

```diff
@@ -113,12 +214,15 @@ def make_gen_state(output, batch, attns, range_, copy_attn=None):
         raise AssertionError("using -copy_attn you need to pass in "
                              "-dynamic_dict during preprocess stage.")
 
+    use_dw = hasattr(batch, "dw")
     return {"output": output,
             "target": batch.tgt[range_[0] + 1: range_[1]],
             "copy_attn": attns.get("copy"),
             "align": None if not copy_attn
             else batch.alignment[range_[0] + 1: range_[1]],
-            "coverage": attns.get("coverage")}
+            "coverage": attns.get("coverage"),
+            "dw": None if not use_dw
+            else batch.dw[range_[0] + 1: range_[1]]}
```

Comment:
Maybe later we could make this subclass specific to get rid of these `if...else...` ?

Reply:
definitely... should opts be propagated to avoid including `copy_attn`, `align` and `dw`?

Reply:
No.  Keep the `*LossCompute` class getting what they really need. We would like to make this like a library that can be used directly from pytorch someday, so it should not care about whether these settings are from command line or other approaches.

Reply:
i'll leave it be for now then!